### ISSUE TITLE
ORBIT-Kick Implementation 

### ISSUE THREAD

**Issue by pbonofig** [ascot4fusion/ascot5#134]
The ORBIT-kick model [1] is a reduced model to encode the wave-particle resonant interactions via changes, or kicks, in the particles' constants of motion ($E$, $P_{\zeta}$, $\mu$). It has proven to be a powerful tool in the EP community and would enhance ASCOT5's capabilities and analysis. 

It is proposed to add the creation of the so-called "kick matrices" which capture the behavior of resonant particles. 

We envision a new flag in the namelist to toggle this calculation/diagnostic, which would include the calculated kick-matrices as new output in the hdf5 file. 

Currently in ORBIT, their are additional namelist parameters to define the kick-matrix calculations. These include the time over which to average (typically a $\delta t$ a little over a mode period), the bounds in ($E$, $P_{\zeta}$, $\mu$) at which to bin, and the bin-spacing or number of bins. These could also be added as namelist options or hard-coded; it depends how much control we want to give the user. 

Since the kick calculations and I/O would constitute moderate additions, a new submodule called "kicks," "pdedp," or something similar would be created. This would contain everything related to the orbit-kicks. 

Since the future of ASCOT5 is looking to transfer to a fully executable python version. This could all be implemented in python. 

# Input
There is no additional input that is needed that ASCOT does not already take. All that is needed are markers and a mode (Technically a mode is not needed. You would just get zero kicks, which is a good test case.)

An implementation to initialize markers uniformly in phase-space would be very useful and can be created in parallel to the kick-matrix development. Likewise, any initialization of markers from a more physical distribution, such as one calculated by NUBEAM, would also be of use. Again, these would be separate functions operating in the marker class. 

Another small, but useful feature, for the above marker initialization would be optional flags to choose specific orbit types. For example, only initializing trapped orbits or trapped and co-passing, etc..  

# Output
The only additional output would be the addition of the kick-matrices to the .h5 file. We would use the "sparse" representation of the matrix which eliminates large regions of phase-space which do not interact with the mode (i.e. zeros). 

A wrapper function would be created for post-processing the kick-matrices into the required UFILE format for input into TRANSP/NUBEAM (for the so-called TRANSP-kick analysis) [2]. 

The output and kick-matrices would also be translated to a new IDS structure, so that it is IMAS compatible and more portable.

Of course, plotting utilities would also be added for plotting the kicks and their features. I already have some of these written in python for ORBIT, so the conversion should be easy. 

# Test Cases
The kick-matrix implementation is solely envisioned for axisymmetric cases since the non-conservation of $P_{\zeta}$ in stellarators makes things more complicated. A large aspect ratio tokamak, such as ITER, with a simple, static, TAE mode is envisioned for benchmarking. 

There is no clear analytical solution for the probability matrices, but their form should follow the linear relationship in [1] such that $\Delta P_{\zeta}/\Delta E \sim n/\omega$. A simple consistency check could be added to verify this. It is also very easy to verify by eye by plotting the kicks. 

A one-to-one comparison to ORBIT would also be done. This should be straightforward: we give the same equilibrium, mode structure, mode amplitude, and marker population, and examine the kick-matrix output. The comparisons can actually be done for many cases, but I imagine a lot of this will be future validation studies and comparisons of reduced models. For now, we will compare to ensure validity of the working model in ASCOT5 and leave higher-fidelity physics testing as continual work in progress. 

# Future Work
There are many adaptations and features currently included in ORBIT which are not envisioned for this project but would make for good future projects. The most notable would be the inclusion of time-dependent mode amplitudes which can capture interesting physics such as sawtooth crashes [3] and chirping behavior [4].

Fine-tuning can also be done to better define and localize resonant particles from those that are near-resonant, e.g. particles that partially slosh back-and-forth on the periphery of the resonance. We have had discussions on various methods for this, but we have not found a clear method to work in all cases. 

# References
[1] M Podesta et al 2014 Plasma Phys. Control. Fusion 56 055003
[2] M Podesta et al 2017 Plasma Phys. Control. Fusion 59 095008
[3] M Podesta et al 2022 Plasma Phys. Control. Fusion 64 025002
[4] A Bierwage et al 2021 Plasma Fusion Res. 16 1403087

**Comment by miekkasarki** [comment]
I've just one comment on the output. It shouldn't be necessary to write the kick matrix to the HDF5 file, since Ascot itself is not going to use it. Instead, we can just write the data directly in the UFILE format (or in IDS in the future).

Of course we would like to visualize the matrix once it has been computed, so in that sense it could be stored in the HDF5 file. However, even for this purpose I'd suggest storing it as a separate HDF5 file or implement tool to read it from UFILE.

Me or @jsjoona can implement the mapping from COM to real space. There are tools that can do this already, but they are stored in `a5py/_nonfunctional` which hints what their current state is. In any case, this mapping is useful for other purposes so we'll develop it separately in separate branch & issue.
